zebra-rpc/src/methods.rs

```diff
@@ -2226,6 +2226,24 @@ where
             extra_coinbase_data,
         );
 
+        // - Validate the block template
+        let network_upgrade = NetworkUpgrade::current(&network, next_block_height);
+
+        let block = proposal_block_from_template(&response, None, network_upgrade)
+            .map_err(|error| ErrorObject::owned(0, error.to_string(), None::<()>))?;
+        let block_bytes = block
+            .zcash_serialize_to_vec()
+            .map_err(|error| ErrorObject::owned(0, error.to_string(), None::<()>))?;
```

Comment:
If the check fails, Zebra should try to remove the invalid transactions. One approach for that could be to map every non-coinbase transaction in the invalid block to a proposal blocks, send all of them to the block verifier, and remove any transactions that were in an invalid proposal block.

It would be nice to factor out block deserialization from `validate_block_proposal()` and add a config flag or ideally a parameter field for skipping this check (if I were a miner, I would want to defer this check until after getting the template, submitting it as a proposal, and seeing that it failed).

Reply:
> (if I were a miner, I would want to defer this check until after getting the template, submitting it as a proposal, and seeing that it failed)

If the deferred check fails, it most likely means there's a fatal bug that caused Zebra to produce an invalid block template, on which the pool has wasted a lot of resources, so I don't think we should allow deferring the check until the block is mined.

Also, the bug has the potential to stall block production entirely, so we should log a huge error and ask for reporting.

Reply:
> I don't think we should allow deferring the check until the block is mined.

Not until the block is mined, I think the default behaviour should be to validate the block, but it would be nice to have a parameter that tells it to skip it so that miners can get the block template and immediately submit it as a proposal ~200ms later while optimistically mining for that fraction of a second.

Reply:
It needs some small adjustments/additions, but the idea is better than what I had in mind. 

> We might want to .. take a different approach altogether

I think it would be even better to do this validation in the mempool by adding another field [here](https://github.com/ZcashFoundation/zebra/blob/main/zebrad/src/components/mempool.rs#L100) for block proposal verification so the mempool can remove transactions from its verified set on its own and ahead of block template generation if they create invalid block proposals, see [this comment](https://github.com/ZcashFoundation/zebra/pull/9525/files#r2101080501).